### PR TITLE
BUG: DataFrame.melt gives unexpected result when column "value" already exists

### LINKED ISSUE #34731: BUG: DataFrame.melt gives unexpected result when column "value" already exists
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
df = pd.DataFrame({'col':list('ABC'),
                   'value':range(10,16,2)})
print(df, '\n')

dfm = df.melt(id_vars='value')

print(dfm)

  col  value
0   A     10
1   B     12
2   C     14 

  value variable value
0     A      col     A
1     B      col     B
2     C      col     C
```

#### Problem description

When the column `value` already exists,  and we set it as index with `id_vars` it copies the `value` column of the melted dataframe and does not set the "old" `value` column of the original dataframe as `id_vars`
#### Expected Output
```python
# note change of column name 'val' instead of 'value'
df = pd.DataFrame({'col':list('ABC'),
                   'val':range(10,16,2)})

dfm = df.melt(id_vars='val')

print(dfm)
   val variable value
0   10      col     A
1   12      col     B
2   14      col     C
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.3.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.5.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 47.1.1.post20200604
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by pizzathief**
- [ x] closes #34731 
- [x ] tests added / passed
- [ X] passes `black pandas`
- [ x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ x] whatsnew entry

### CODE REVIEW

**Comment by pizzathief** [comment]
> Linting .py code
> 16
> ##[error]./pandas/core/apply.py:9:1:F401:'pandas._libs.reduction as libreduction' imported but unused
> 

should I remove this line as part of this PR? "libreduction" isn't mentioned anywhere else in the file pandas/core/apply.py , but on the other hand, its out of scope

**Comment by WillAyd** [comment]
Can you merge master to fix CI issue?

File: pandas/tests/reshape/test_melt.py [hunk 1]
```diff
@@ -1014,3 +1016,16 @@ def test_col_substring_of_stubname(self):
...
         result = pd.wide_to_long(wide_df, stubnames="PA", i=["node_id", "A"], j="time")
         tm.assert_frame_equal(result, expected)
+
+    def test_warn_of_column_name_value(self):
+        # GH34731
+        # raise a warning if the resultant value column name matches
+        # a name in the dataframe already (default name is "value")
+        df = pd.DataFrame({"col": list("ABC"), "value": range(10, 16, 2)})
+
+        with warnings.catch_warnings(record=True) as w:
```
Comment (dsaxton) [inline pandas/tests/reshape/test_melt.py:1026]: Can you use tm.assert_produces_warning here (e.g., as in L580 of this file)?

File: pandas/core/reshape/melt.py [hunk 2]
```diff
@@ -40,6 +41,14 @@ def melt(
     else:
         cols = list(frame.columns)
 
+    if value_name in frame.columns:
+        warnings.warn(
+            'The value column name "%s" conflicts with an existing '
```
Comment (WillAyd) [inline pandas/core/reshape/melt.py:46]: Can you use f-string syntax? We don't use the Py2 syntax any more

File: pandas/core/reshape/melt.py [hunk 3]
```diff
@@ -40,6 +41,14 @@ def melt(
...
+            "column in the dataframe." % (value_name),
+            DeprecationWarning,
```
Comment (WillAyd) [inline pandas/core/reshape/melt.py:48]: Can you change this to a FutureWarning?

File: doc/source/whatsnew/v1.1.0.rst [hunk 4]
```diff
@@ -1105,6 +1105,7 @@ Reshaping
 - Fixed bug in :func:`melt` where melting MultiIndex columns with ``col_level`` > 0 would raise a ``KeyError`` on ``id_vars`` (:issue:`34129`)
 - Bug in :meth:`Series.where` with an empty Series and empty ``cond`` having non-bool dtype (:issue:`34592`)
 - Fixed regression where :meth:`DataFrame.apply` would raise ``ValueError`` for elements whth ``S`` dtype (:issue:`34529`)
+- Issue warning if value column name already exists when using :meth:`DataFrame.melt` (:issue:`34731`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.1.0.rst:1108]: Can you move this to the Deprecations section?

File: pandas/core/reshape/melt.py [hunk 5]
```diff
@@ -40,6 +41,14 @@ def melt(
     else:
         cols = list(frame.columns)
 
+    if value_name in frame.columns:
+        warnings.warn(
+            'The value column name "%s" conflicts with an existing '
+            "column in the dataframe." % (value_name),
```
Comment (WillAyd) [inline pandas/core/reshape/melt.py:47]: Generally we want to give users insight as to what they need to do before the deprecated behavior gets changed - can you try to clarify the message a bit?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.1.0.rst [hunk 6]
```diff
@@ -816,6 +816,7 @@ Deprecations
 - :meth:`util.testing.assert_almost_equal` now accepts both relative and absolute
   precision through the ``rtol``, and ``atol`` parameters, thus deprecating the
   ``check_less_precise`` parameter. (:issue:`13357`).
+- Issue warning if value column name already exists when using :meth:`DataFrame.melt` (:issue:`34731`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:819]: can you start with `DataFrame.melt` (like the other notes above), and also say that this is deprecated and removed in future version.

File: pandas/core/reshape/melt.py [hunk 7]
```diff
@@ -40,6 +41,16 @@ def melt(
     else:
         cols = list(frame.columns)
 
+    if value_name in frame.columns:
+        warnings.warn(
+            "This dataframe has a column name that matches the value column "
```
Comment (jreback) [inline pandas/core/reshape/melt.py:46]: put `value` in single quotes as this is an argument that the user passed. (and this should be `value_name`)

File: pandas/core/reshape/melt.py [hunk 8]
```diff
@@ -40,6 +41,16 @@ def melt(
...
+            f'name of the resultant melted dataframe (That being "{value_name})". '
```
Comment (jreback) [inline pandas/core/reshape/melt.py:47]: resultant -> resuling DataFrame. you don't need the parenthesized expression.

File: pandas/core/reshape/melt.py [hunk 9]
```diff
@@ -40,6 +41,16 @@ def melt(
...
+            "In the future this will raise an error, please set the value_name "
```
Comment (jreback) [inline pandas/core/reshape/melt.py:48]: `value_name` in single quotes

File: pandas/tests/reshape/test_melt.py [hunk 10]
```diff
@@ -1014,3 +1014,12 @@ def test_col_substring_of_stubname(self):
...
         tm.assert_frame_equal(result, expected)
+
+    def test_warn_of_column_name_value(self):
+        # GH34731
+        # raise a warning if the resultant value column name matches
+        # a name in the dataframe already (default name is "value")
+        df = pd.DataFrame({"col": list("ABC"), "value": range(10, 16, 2)})
+
+        with tm.assert_produces_warning(FutureWarning):
+            dfm = df.melt(id_vars="value")  # noqa F841
```
Comment (jreback) [inline pandas/tests/reshape/test_melt.py:1025]: can you create an expected value and use assert_frame_equal

also instead of dfm, call this result.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @pizzathief

**Comment by jbrockmendel** [comment]
@pizzathief want to make a PR to enforce this deprecation?

File: pandas/core/reshape/melt.py [hunk 11]
```diff
@@ -40,6 +41,16 @@ def melt(
     else:
         cols = list(frame.columns)
 
+    if value_name in frame.columns:
+        warnings.warn(
+            "This dataframe has a column name that matches the 'value_name' column "
+            "name of the resultiing Dataframe. "
```
Comment (samukweku) [inline pandas/core/reshape/melt.py:47]: the spelling here should be `resulting`, not `resultiing`

**Pull request merged**